presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -776,22 +788,80 @@ public Page getNextPage()
                 if (qualifyingSet == null) {
                     qualifyingSet = new QualifyingSet();
                 }
-                qualifyingSet.setRange((int) currentGroupRowCount);
+                qualifyingSet.setRange(0, Math.min(ariaBatchRows, (int) currentGroupRowCount));
                 reader.setQualifyingSets(qualifyingSet, null);
+                if ((currentRowGroup & 0x3) == 0) {
+                    // Decay row size stats every 4 groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                }
                 if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
```

Comment:
`(currentRowGroup & 0x3) != 0` - isn't this true for every 3 groups output of 4? it would be clearer to combine all actions happening for every 4th row group:

```
                if ((currentRowGroup & 0x3) == 0) {
                    // Decay row size stats and reconsider filter order every 4 groups.
                    numAriaRows /= 2;
                    numAriaBytes /= 2;
                    if (reorderFilters && currentRowGroup > 0) {
                        reader.maybeReorderFilters();
                    }
                }
```